packages/node-sdk/src/rate-limiter.ts

```diff
@@ -0,0 +1,63 @@
+import { clearInterval } from "timers";
+
+import { ok } from "./utils";
+
+/**
+ * Creates a new rate limiter.
+ *
+ * @typeparam TKey - The type of the key.
+ * @param windowSizeMs - The length of the time window in milliseconds.
+ *
+ * @returns The rate limiter.
+ **/
+export function newRateLimiter(windowSizeMs: number) {
+  ok(
+    typeof windowSizeMs == "number" && windowSizeMs > 0,
+    "windowSizeMs must be greater than 0",
+  );
+
+  let lastAllowedTimestampsByKey: { [key: string]: number } = {};
+  let clearIntervalId: NodeJS.Timeout | undefined;
+
+  function clear(all: boolean): void {
+    if (clearIntervalId) {
+      clearInterval(clearIntervalId);
+      clearIntervalId = undefined;
+    }
+
+    if (all) {
+      lastAllowedTimestampsByKey = {};
+    } else {
+      const expiredAfter = Date.now() - windowSizeMs;
+      const keys = Object.keys(lastAllowedTimestampsByKey);
+
+      for (const key in keys) {
+        const lastAllowedTimestamp = lastAllowedTimestampsByKey[key];
+
+        if (lastAllowedTimestamp < expiredAfter) {
+          delete lastAllowedTimestampsByKey[key];
+        }
+      }
+    }
+  }
+
+  function isAllowed(key: string): boolean {
+    clearIntervalId =
+      clearIntervalId || setInterval(() => clear(false), windowSizeMs);
```

Comment:
I think could be `setTimeout` since we're not using the intercal functionality?

Reply:
Actually wee need that, but the clearing was too aggressive, should only kill the timeout if there's no smoke keys left in the cache. I rearranged the code.

If we keep setting a new interval each time `isAllowed` is invoked then we risk into never actually running the timer since it would be renewed every time a check is done (which could be less than once a minute).

We could also invoke `clear(false)` in the `isAllowed` and only start a timer if needed, but potato/potatoh :)